TypeLookupService doesn't support SCCs

jrw972 opened the pull request:
Problem
-------

The TypeLookupService doesn't support strongly connected components (SCCs) correctly.

Solution
--------

* Add the SCCId to dependencies when necessary (fixes `getTypeDependencies`).
* Update `getTypes` to consolidate type identifiers and send all components of an SCC.

jrw972 commented:
See https://github.com/OpenDDS/OpenDDS/pull/4938

simpsont-oci commented:
I haven't had a time to do a full review, but I've built this branch locally and tested it against the monitor & (unmanaged) monitor test app and unfortunately the monitor still crashes. I've got a bunch of debugging changes stashed somewhere, I'll try to salvage those and see if I can figure out what's not getting requested / sent (assuming it's even crashing in the same spot).

sonndinh commented on dds/DCPS/XTypes/TypeObject.cpp:
```diff
@@ -471,616 +471,719 @@ const TypeMap TypeMapBuilder::EmptyMap;
+
+  TypeIdentifier make_scc_id(const TypeIdentifier& tid)
+  {
+    if (tid.kind() != TI_STRONGLY_CONNECTED_COMPONENT) {
+      return tid;
```
Should this return an error instead? It is returning the input type which isn't an SCC which can be misleading given the function name `make_scc_id`.

jrw972 replied:
The function does exactly what I want it to do, return the SCCId if it's a component of an SCC or the given type identifier if not.  Would `make_scc_id_or_default` be better?

sonndinh replied:
Yeah, I think so.

sonndinh commented on dds/DCPS/XTypes/TypeObject.cpp:
```diff
@@ -471,616 +471,719 @@ const TypeMap TypeMapBuilder::EmptyMap;
+      {
+        // An SCCId (identifies the entire component.)
+        TypeIdentifier scc_id = make_scc_id(type_identifier);
+        sccs.insert(scc_id);
```
The returned `scc_id` has `scc_index` value of 0. I think the SCC index starts from 1?

jrw972 commented on dds/DCPS/XTypes/TypeLookupService.cpp:
```diff
@@ -57,11 +57,27 @@ TypeLookupService::~TypeLookupService()
+  }
+
+  ACE_GUARD(ACE_Thread_Mutex, g, mutex_);
+  for (TypeIdentifierSet::const_iterator pos = tids.begin(), limit = tids.end(); pos != limit; ++pos) {
+    const TypeIdentifier& tid = *pos;
+    TypeMap::const_iterator pos2 = type_map_.find(tid);
+    if (pos2 != type_map_.end()) {
+      types.append(TypeIdentifierTypeObjectPair(tid, pos2->second));
```
It's possible but I don't think that's the problem based on what @simpsont-oci has said.

simpsont-oci commented:
> I haven't had a time to do a full review, but I've built this branch locally and tested it against the monitor & (unmanaged) monitor test app and unfortunately the monitor still crashes. I've got a bunch of debugging changes stashed somewhere, I'll try to salvage those and see if I can figure out what's not getting requested / sent (assuming it's even crashing in the same spot).

I was able to restore my debugging output... it looks like what's happening is that the SCC type ids are never getting sent or requested. I see an error message on the testapp in `populate_dependencies_i` which makes it look like the zero-index SCC type identifiers can't be found in type_map and thus can't be serialized (can't create TypeIdentifierWithSize), and I think that's generally true for our implementation, only the individual SCC components get added to the type_map, but maybe that needs to change?

jrw972 commented:
I added a test for an SCC topic type.
For the IDL I'm using, it should have 5 types in the SCC and it does.
They are indexed from 1 to 5 and they are inserted into the type map.

`getTypeDependencies` should return the index 0 SCCId (see below) when queried with any non-index SCCId or the index 0 SCCId.
I updated the calling code in OpenDDS to the index 0 SCCId based on the spec.
`getTypes` should return type objects for all of the components of an SCCId.

To me, the spec is vague about returning the type in its dependency list.
If you query for the dependencies of X should you
1. Always get back X
2. Never get back X
3. Only get back the index 0 SCCId of X if X is part of an SCC

Right now, we are doing 3.

simpsont-oci commented:
Poked around with the latest changes a bit more, and it looks like what's happening for the monitor is this:
 - Complex topic type is not an SCC, but depends on one.
 - Request for that topic's dependencies seems to correctly compute the the dependency, but when it goes calculate the response size in `TypeSupportImpl::populate_dependencies_i`, the the 0-index SCC doesn't exist in the type map so we get an error message and the and the SCC dependency gets omitted from the response.
 - If we make a special check for 0-index SCCs in `TypeSupportImpl::populate_dependencies_i`, then a similar thing happens in `TypeLookupService::get_type_dependencies_i` when it tries to actually serialize the response.
 - If we also make a special check for 0-index SCCs in `TypeLookupService::get_type_dependencies_i`, it then looks like the monitor is able to receive the 0-index SCC and then make a request for it, getting all the individual components and not crash because of a null descriptor when trying to subscribe to the complex topic.

Even with those changes, however, I also see a bunch of these warnings, which I don't remember seeing before, but I'm not sure if these changes could have caused them:
```
(1749885|1749885) NOTICE: DynamicDataImpl::get_value_from_struct: trying to get test::BasicTypes.u8 id 0 kind uint8 (uint8) as an invalid kind uint16
(1749885|1749885) NOTICE: DynamicDataImpl::get_value_from_struct: trying to get test::BasicTypes.i8 id 1 kind int8 (int8) as an invalid kind int16
(1749885|1749885) NOTICE: DynamicDataImpl::get_value_from_struct: trying to get test::BasicTypes.u8 id 0 kind uint8 (uint8) as an invalid kind uint16
(1749885|1749885) NOTICE: DynamicDataImpl::get_value_from_struct: trying to get test::BasicTypes.i8 id 1 kind int8 (int8) as an invalid kind int16
```

simpsont-oci commented:
> Even with those changes, however, I also see a bunch of these warnings, which I don't remember seeing before, but I'm not sure if these changes could have caused them:

Just double checked against master, and these warnings/errors are there too, just hidden in the log window. The mapping seems odd in the monitor (TK_UINT8 becomes tk_ushort?) so it's likely a bug there.

jrw972 commented:
> Poked around with the latest changes a bit more, and it looks like what's happening for the monitor is this:
> 
> * Complex topic type is not an SCC, but depends on one.
> * Request for that topic's dependencies seems to correctly compute the the dependency, but when it goes calculate the response size in `TypeSupportImpl::populate_dependencies_i`, the the 0-index SCC doesn't exist in the type map so we get an error message and the and the SCC dependency gets omitted from the response.
> * If we make a special check for 0-index SCCs in `TypeSupportImpl::populate_dependencies_i`, then a similar thing happens in `TypeLookupService::get_type_dependencies_i` when it tries to actually serialize the response.
> * If we also make a special check for 0-index SCCs in `TypeLookupService::get_type_dependencies_i`, it then looks like the monitor is able to receive the 0-index SCC and then make a request for it, getting all the individual components and not crash because of a null descriptor when trying to subscribe to the complex topic.
> 
> Even with those changes, however, I also see a bunch of these warnings, which I don't remember seeing before, but I'm not sure if these changes could have caused them:
> 
> ```
> (1749885|1749885) NOTICE: DynamicDataImpl::get_value_from_struct: trying to get test::BasicTypes.u8 id 0 kind uint8 (uint8) as an invalid kind uint16
> (1749885|1749885) NOTICE: DynamicDataImpl::get_value_from_struct: trying to get test::BasicTypes.i8 id 1 kind int8 (int8) as an invalid kind int16
> (1749885|1749885) NOTICE: DynamicDataImpl::get_value_from_struct: trying to get test::BasicTypes.u8 id 0 kind uint8 (uint8) as an invalid kind uint16
> (1749885|1749885) NOTICE: DynamicDataImpl::get_value_from_struct: trying to get test::BasicTypes.i8 id 1 kind int8 (int8) as an invalid kind int16
> ```

Excellent detective work.

So, it seems like the missing piece is the size of SCCId(0).
And the spec is not going to help us.
I see three possible options:
1. Let the size be 0.
2. The sum of the serialized sizes of the components.
3. The serialized size of a sequence of type objects of the components.  (This is probably the one implied by the spec.)

simpsont-oci commented:
> > Poked around with the latest changes a bit more, and it looks like what's happening for the monitor is this:
> > 
> > * Complex topic type is not an SCC, but depends on one.
> > * Request for that topic's dependencies seems to correctly compute the the dependency, but when it goes calculate the response size in `TypeSupportImpl::populate_dependencies_i`, the the 0-index SCC doesn't exist in the type map so we get an error message and the and the SCC dependency gets omitted from the response.
> > * If we make a special check for 0-index SCCs in `TypeSupportImpl::populate_dependencies_i`, then a similar thing happens in `TypeLookupService::get_type_dependencies_i` when it tries to actually serialize the response.
> > * If we also make a special check for 0-index SCCs in `TypeLookupService::get_type_dependencies_i`, it then looks like the monitor is able to receive the 0-index SCC and then make a request for it, getting all the individual components and not crash because of a null descriptor when trying to subscribe to the complex topic.
> > 
> > Even with those changes, however, I also see a bunch of these warnings, which I don't remember seeing before, but I'm not sure if these changes could have caused them:
> > ```
> > (1749885|1749885) NOTICE: DynamicDataImpl::get_value_from_struct: trying to get test::BasicTypes.u8 id 0 kind uint8 (uint8) as an invalid kind uint16
> > (1749885|1749885) NOTICE: DynamicDataImpl::get_value_from_struct: trying to get test::BasicTypes.i8 id 1 kind int8 (int8) as an invalid kind int16
> > (1749885|1749885) NOTICE: DynamicDataImpl::get_value_from_struct: trying to get test::BasicTypes.u8 id 0 kind uint8 (uint8) as an invalid kind uint16
> > (1749885|1749885) NOTICE: DynamicDataImpl::get_value_from_struct: trying to get test::BasicTypes.i8 id 1 kind int8 (int8) as an invalid kind int16
> > ```
> 
> Excellent detective work.
> 
> So, it seems like the missing piece is the size of SCCId(0). And the spec is not going to help us. I see three possible options:
> 
>     1. Let the size be 0.
> 
>     2. The sum of the serialized sizes of the components.
> 
>     3. The serialized size of a sequence of type objects of the components.  (This is probably the one implied by the spec.)
> 
> 
> https://issues.omg.org/browse/DDSXTY14-87

I'm not sure I'm following... wouldn't it just be the size of the 0-index SCC TypeIdentifier?

I take it from the 2nd-to-last paragraph of section 7.6.3.2.1 that it's fine to just return the 0-index SCC TypeIdentifier in the list of dependencies for the queried type/s, and then from that single 0-index SCC TypeIdentifier, the interested participant can follow up with a type request containing all of the associated component SCC TypeIdentifiers, at which point it will get the actual TypeObjects for them.

simpsont-oci commented:
For example, the code I added to `populate_dependencies_i` looks like this (ignore the DEBUG statements):
```
   for (; it != dependencies.end(); ++it) {
-    XTypes::TypeMap::const_iterator iter = type_map.find(*it);
-    if (iter != type_map.end()) {
-      const size_t tobj_size = serialized_size(XTypes::get_typeobject_encoding(), iter->second);
+    ACE_DEBUG((LM_DEBUG, ACE_TEXT("TypeSupportImpl::populate_dependencies_i - found dependency %C\n"), it->to_string().c_str()));
+    if (it->kind() == XTypes::TI_STRONGLY_CONNECTED_COMPONENT && it->sc_component_id().scc_index == 0) {
+      const size_t tobj_size = serialized_size(XTypes::get_typeobject_encoding(), *it);
       XTypes::TypeIdentifierWithSize ti_with_size(*it, static_cast<ACE_CDR::ULong>(tobj_size));
+      ACE_DEBUG((LM_DEBUG, ACE_TEXT("TypeSupportImpl::populate_dependencies_i - found SCC dependency with calculated size of %d\n"), tobj_size));
       deps_with_size.append(ti_with_size);
-    } else if (XTypes::has_type_object(*it)) {
-      log_ti_not_found("populate_dependencies_i", name(), *it);
+    } else {
+      XTypes::TypeMap::const_iterator iter = type_map.find(*it);
+      if (iter != type_map.end()) {
+        const size_t tobj_size = serialized_size(XTypes::get_typeobject_encoding(), iter->second);
+        XTypes::TypeIdentifierWithSize ti_with_size(*it, static_cast<ACE_CDR::ULong>(tobj_size));
+        deps_with_size.append(ti_with_size);
+      } else if (XTypes::has_type_object(*it)) {
+        log_ti_not_found("populate_dependencies_i", name(), *it);
+      }
     }
```

and the code I added to `get_type_dependencies_i` looks like this (again, ignore DEBUG stuff):
```
   for (unsigned i = 0; iter != tmp.end(); ++i, ++iter) {
-    const TypeMap::const_iterator tobj_it = type_map_.find(*iter);
-    if (tobj_it != type_map_.end()) {
+    if (iter->kind() == TI_STRONGLY_CONNECTED_COMPONENT && iter->sc_component_id().scc_index == 0) {
       dependencies[i].type_id = *iter;
-      const size_t sz = DCPS::serialized_size(get_typeobject_encoding(), tobj_it->second);
+      const size_t sz = serialized_size(get_typeobject_encoding(), *iter);
+      ACE_DEBUG((LM_DEBUG, ACE_TEXT("TypeLookupService::get_type_dependencies_i - found SCC dependency with calculated size of %d\n"), sz));
       dependencies[i].typeobject_serialized_size = static_cast<unsigned>(sz);
+    } else { 
+      const TypeMap::const_iterator tobj_it = type_map_.find(*iter);
+      if (tobj_it != type_map_.end()) {
+        dependencies[i].type_id = *iter;
+        const size_t sz = DCPS::serialized_size(get_typeobject_encoding(), tobj_it->second);
+        dependencies[i].typeobject_serialized_size = static_cast<unsigned>(sz);
+      }
     }
   }
```

jrw972 commented:
> > > Poked around with the latest changes a bit more, and it looks like what's happening for the monitor is this:
> > > 
> > > * Complex topic type is not an SCC, but depends on one.
> > > * Request for that topic's dependencies seems to correctly compute the the dependency, but when it goes calculate the response size in `TypeSupportImpl::populate_dependencies_i`, the the 0-index SCC doesn't exist in the type map so we get an error message and the and the SCC dependency gets omitted from the response.
> > > * If we make a special check for 0-index SCCs in `TypeSupportImpl::populate_dependencies_i`, then a similar thing happens in `TypeLookupService::get_type_dependencies_i` when it tries to actually serialize the response.
> > > * If we also make a special check for 0-index SCCs in `TypeLookupService::get_type_dependencies_i`, it then looks like the monitor is able to receive the 0-index SCC and then make a request for it, getting all the individual components and not crash because of a null descriptor when trying to subscribe to the complex topic.
> > > 
> > > Even with those changes, however, I also see a bunch of these warnings, which I don't remember seeing before, but I'm not sure if these changes could have caused them:
> > > ```
> > > (1749885|1749885) NOTICE: DynamicDataImpl::get_value_from_struct: trying to get test::BasicTypes.u8 id 0 kind uint8 (uint8) as an invalid kind uint16
> > > (1749885|1749885) NOTICE: DynamicDataImpl::get_value_from_struct: trying to get test::BasicTypes.i8 id 1 kind int8 (int8) as an invalid kind int16
> > > (1749885|1749885) NOTICE: DynamicDataImpl::get_value_from_struct: trying to get test::BasicTypes.u8 id 0 kind uint8 (uint8) as an invalid kind uint16
> > > (1749885|1749885) NOTICE: DynamicDataImpl::get_value_from_struct: trying to get test::BasicTypes.i8 id 1 kind int8 (int8) as an invalid kind int16
> > > ```
> > 
> > 
> > Excellent detective work.
> > So, it seems like the missing piece is the size of SCCId(0). And the spec is not going to help us. I see three possible options:
> > ```
> > 1. Let the size be 0.
> > 
> > 2. The sum of the serialized sizes of the components.
> > 
> > 3. The serialized size of a sequence of type objects of the components.  (This is probably the one implied by the spec.)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > https://issues.omg.org/browse/DDSXTY14-87
> 
> I'm not sure I'm following... wouldn't it just be the size of the 0-index SCC TypeIdentifier?
> 
> I take it from the 2nd-to-last paragraph of section 7.6.3.2.1 that it's fine to just return the 0-index SCC TypeIdentifier in the list of dependencies for the queried type/s, and then from that single 0-index SCC TypeIdentifier, the interested participant can follow up with a type request containing all of the associated component SCC TypeIdentifiers, at which point it will get the actual TypeObjects for them.

There is no TypeObject corresponding to the SCCIdentifier.  The closest there is the sequence of component type objects from the hashing algorithm.
7.6.3.2.1 and elsewhere says you can only use the SCCIdentifier, except when returning `getTypes` at which point you have to use the identifiers for the individual components.

To summarize:

`getTypeDependencies`
* in: TypeIdentifier (minimal/complete) or SCCIdentifier (index 0).
* out: TypeIdentifierWithSize (same as above) <-- This is where the spec lacks clarity.

`getTypes`
* in: TypeIdentifier (minimal/complete) or SCCIdentifier (index 0).
* out: TypeIdentifier and TypeObject (SCCComponentIds must have index != 0 and all must be present)

simpsont-oci commented:
> There is no TypeObject corresponding to the SCCIdentifier. The closest there is the sequence of component type objects from the hashing algorithm. 7.6.3.2.1 and elsewhere says you can only use the SCCIdentifier, except when returning `getTypes` at which point you have to use the identifiers for the individual components.
> 
> To summarize:
> 
> `getTypeDependencies`
> 
>     * in: TypeIdentifier (minimal/complete) or SCCIdentifier (index 0).
> 
>     * out: TypeIdentifierWithSize (same as above) <-- This is where the spec lacks clarity.
> 
> 
> `getTypes`
> 
>     * in: TypeIdentifier (minimal/complete) or SCCIdentifier (index 0).
> 
>     * out: TypeIdentifier and TypeObject (SCCComponentIds must have index != 0 and all must be present)

Ah, I see now, I didn't realize the "Size" part of TypeIdentifiersWithSize was supposed to be the size of the TypeObject rather than the TypeIdentifier. Which makes more sense now that I think about it. In the code I shared I think I'm just returning the size of the serialized TypeIdentifier.

simpsont-oci commented on dds/DCPS/XTypes/TypeLookupService.cpp:
```diff
@@ -103,27 +123,17 @@ void TypeLookupService::get_type_dependencies(const TypeIdentifierSeq& type_ids,
     if (it != type_dependencies_map_.end()) {
       for (unsigned j = 0; j < it->second.length(); ++j) {
-        tmp.insert(it->second[j].type_id);
+        const size_t idx = dependencies.length();
```
```suggestion
        const auto idx = dependencies.length();
```

simpsont-oci replied:
signed / unsigned warning

simpsont-oci commented on dds/DCPS/XTypes/TypeObject.cpp:
```diff
@@ -997,24 +1012,32 @@ namespace {
+                              TypeIdentifierSet& dependencies)
   {
-    // Do nothing.
+    // An SCCId (identifies the entire component.)
```
If I have a topic type T which (eventually) depends on an SCC, and I request the dependencies of just type T... doesn't this code mean I'll _always_ get all the dependencies for the whole SCC (even if I haven't asked explicitly asked for it yet?). Is that what's intended / desired?

jrw972 replied:
That is the intended behavior.  If you ask for the dependencies of T, you should get all of the dependencies of T.
